File: app/src/main/java/com/ravisharma/playbackmusic/database/PlaylistDatabase.java
package com.ravisharma.playbackmusic.database;

import android.content.Context;
import android.os.AsyncTask;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.sqlite.db.SupportSQLiteDatabase;

import com.ravisharma.playbackmusic.R;
import com.ravisharma.playbackmusic.database.model.DatabaseSetup;
import com.ravisharma.playbackmusic.database.model.QueueSongs;
import com.ravisharma.playbackmusic.database.model.ShuffleSongs;
import com.ravisharma.playbackmusic.model.Playlist;
import com.ravisharma.playbackmusic.model.Song;
import com.ravisharma.playbackmusic.prefrences.PrefManager;
import com.ravisharma.playbackmusic.prefrences.TinyDB;

import java.lang.ref.WeakReference;
import java.util.ArrayList;

@Database(entities = {Playlist.class, QueueSongs.class, ShuffleSongs.class, DatabaseSetup.class}, version = 1)
public abstract class PlaylistDatabase extends RoomDatabase {

    public abstract PlaylistDao playlistDao();

    public abstract QueueSongsDao queueSongsDao();

    public abstract ShuffleSongsDao shuffleSongsDao();

    public abstract SetupDao setupDao();

    private static PlaylistDatabase instance = null;
    private static Context context;

    public static synchronized PlaylistDatabase getInstance(Context context) {
        if (instance == null) {
            PlaylistDatabase.context = context;
            instance = Room.databaseBuilder(context,
                    PlaylistDatabase.class,
                    "PlaylistDB")
                    .addCallback(callback)
                    .allowMainThreadQueries()
                    .build();
        }
        return instance;
    }

    private static RoomDatabase.Callback callback = new Callback() {
        @Override
        public void onCreate(@NonNull SupportSQLiteDatabase db) {
            super.onCreate(db);
            transferSongsInDB();
        }

        @Override
        public void onOpen(@NonNull SupportSQLiteDatabase db) {
            super.onOpen(db);
            String queryString = "UPDATE sqlite_sequence SET seq=0 WHERE name='queueSongs'";
            String queryString2 = "UPDATE sqlite_sequence SET seq=0 WHERE name='shuffledSongs'";
            db.execSQL(queryString);
            db.execSQL(queryString2);
        }
    };

    private static void transferSongsInDB() {
        new TransferSongAsync(PlaylistDatabase.context).execute();
    }

    private static class TransferSongAsync extends AsyncTask<Void, Void, Void> {
        TinyDB tinydb;
        PrefManager p;
        String fav;

        TransferSongAsync(Context context) {
            tinydb = new TinyDB(context);
            p = new PrefManager(context);
            fav = context.getString(R.string.favTracks);
        }

        @Override
        protected Void doInBackground(Void... voids) {
            ArrayList<String> list = new ArrayList<>();
            list.add(fav);
            list.addAll(p.getAllPlaylist());

            for (String pName : list) {
                ArrayList<Song> songList = new ArrayList<>();

                songList.addAll(tinydb.getListObject(pName, Song.class));

                for (Song song : songList) {
                    Playlist playlist = new Playlist(0, pName, song);

                    instance.playlistDao().addSong(playlist);
                }
            }

            ArrayList<Song> shuffledSongs = new ArrayList<>();
            shuffledSongs.addAll(tinydb.getListObject(context.getString(R.string.Songs), Song.class));
            for (Song song : shuffledSongs) {
                ShuffleSongs s1 = new ShuffleSongs(0, song);
                instance.shuffleSongsDao().addSong(s1);
            }

            ArrayList<Song> normalSongs = new ArrayList<>();
            normalSongs.addAll(tinydb.getListObject(context.getString(R.string.NormalSongs), Song.class));
            for (Song song : normalSongs) {
                QueueSongs s1 = new QueueSongs(0, song);
                instance.queueSongsDao().addSongs(s1);
            }

            DatabaseSetup setup = new DatabaseSetup(1,true);

            instance.setupDao().setUp(setup);

            return null;
        }
    }
}

File: app/src/main/java/com/ravisharma/playbackmusic/database/PlaylistDatabase.kt
package com.ravisharma.playbackmusic.database

import android.content.Context
import androidx.lifecycle.map
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import com.ravisharma.playbackmusic.R
import com.ravisharma.playbackmusic.database.dao.LastPlayedDao
import com.ravisharma.playbackmusic.database.dao.MostPlayedDao
import com.ravisharma.playbackmusic.database.dao.PlaylistDao
import com.ravisharma.playbackmusic.database.model.LastPlayed
import com.ravisharma.playbackmusic.database.model.MostPlayed
import com.ravisharma.playbackmusic.model.Playlist
import com.ravisharma.playbackmusic.model.Song
import com.ravisharma.playbackmusic.prefrences.PrefManager
import com.ravisharma.playbackmusic.prefrences.TinyDB
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import java.util.*
import kotlin.collections.ArrayList

@Database(entities = [Playlist::class, MostPlayed::class, LastPlayed::class], version = 3)
abstract class PlaylistDatabase : RoomDatabase() {

    abstract fun playlistDao(): PlaylistDao
    abstract fun lastPlayedDao(): LastPlayedDao
    abstract fun mostPlayedDao(): MostPlayedDao

    companion object {
        @Volatile
        private var instance: PlaylistDatabase? = null
        private var context: Context? = null

        @Synchronized
        fun getInstance(context: Context?): PlaylistDatabase {
            if (instance == null) {
                Companion.context = context
                instance = Room.databaseBuilder(
                    context!!, PlaylistDatabase::class.java, "PlaylistDB"
                ).addCallback(callback).addMigrations(MIGRATION_1_2, MIGRATION_2_3)
                    .allowMainThreadQueries().build()
            }
            return instance as PlaylistDatabase
        }

        private val MIGRATION_1_2: Migration = object : Migration(1, 2) {
            override fun migrate(database: SupportSQLiteDatabase) {
                database.execSQL("DROP TABLE IF EXISTS setupTable")
                database.execSQL("DROP TABLE IF EXISTS queueSongs")
                database.execSQL("DROP TABLE IF EXISTS shuffledSongs")

                database.execSQL("CREATE TABLE `playlistTable_tmp` (`playlistId` INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, `playlistName` TEXT NOT NULL, `id` INTEGER NOT NULL, `title` TEXT, `artist` TEXT, `art` TEXT, `duration` INTEGER NOT NULL, `data` TEXT, `dateModified` TEXT, `album` TEXT, `composer` TEXT)")
                database.execSQL("INSERT INTO `playlistTable_tmp` (playlistId, playlistName, id, title, artist, art, duration, data, dateModified, album, composer) SELECT playlistId, playlistName, id, title, artist, art, duration, data, dateModified, album, composer FROM playlistTable")
                database.execSQL("DROP TABLE playlistTable")
                database.execSQL("ALTER TABLE playlistTable_tmp RENAME TO playlistTable")
                database.execSQL("CREATE TABLE IF NOT EXISTS `lastPlayed` (`playedId` INTEGER PRIMARY KEY AUTOINCREMENT, `id` INTEGER NOT NULL, `title` TEXT, `artist` TEXT, `art` TEXT, `duration` INTEGER NOT NULL, `data` TEXT, `dateModified` TEXT, `album` TEXT, `composer` TEXT)")
                database.execSQL("CREATE TABLE IF NOT EXISTS `mostPlayed` (`playedId` INTEGER PRIMARY KEY AUTOINCREMENT, `playedCount` INTEGER NOT NULL, `id` INTEGER NOT NULL, `title` TEXT, `artist` TEXT, `art` TEXT, `duration` INTEGER NOT NULL, `data` TEXT, `dateModified` TEXT, `album` TEXT, `composer` TEXT)")
            }
        }
        private val MIGRATION_2_3: Migration = object : Migration(2, 3) {
            override fun migrate(database: SupportSQLiteDatabase) {
                database.execSQL("DROP TABLE IF EXISTS setupTable")
                database.execSQL("DROP TABLE IF EXISTS queueSongs")
                database.execSQL("DROP TABLE IF EXISTS shuffledSongs")

                database.execSQL("CREATE TABLE `playlistTable_tmp` (`playlistId` INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, `playlistName` TEXT NOT NULL, `id` INTEGER NOT NULL, `title` TEXT, `artist` TEXT, `art` TEXT, `duration` INTEGER NOT NULL, `data` TEXT, `dateModified` TEXT, `album` TEXT, `composer` TEXT)")
                database.execSQL("INSERT INTO `playlistTable_tmp` (playlistId, playlistName, id, title, artist, art, duration, data, dateModified, album, composer) SELECT playlistId, playlistName, id, title, artist, art, duration, data, dateModified, album, composer FROM playlistTable")
                database.execSQL("DROP TABLE playlistTable")
                database.execSQL("ALTER TABLE playlistTable_tmp RENAME TO playlistTable")
                database.execSQL("CREATE TABLE IF NOT EXISTS `lastPlayed` (`playedId` INTEGER PRIMARY KEY AUTOINCREMENT, `id` INTEGER NOT NULL, `title` TEXT, `artist` TEXT, `art` TEXT, `duration` INTEGER NOT NULL, `data` TEXT, `dateModified` TEXT, `album` TEXT, `composer` TEXT)")
                database.execSQL("CREATE TABLE IF NOT EXISTS `mostPlayed` (`playedId` INTEGER PRIMARY KEY AUTOINCREMENT, `playedCount` INTEGER NOT NULL, `id` INTEGER NOT NULL, `title` TEXT, `artist` TEXT, `art` TEXT, `duration` INTEGER NOT NULL, `data` TEXT, `dateModified` TEXT, `album` TEXT, `composer` TEXT)")
            }
        }

        private val callback: Callback = object : Callback() {
            override fun onCreate(db: SupportSQLiteDatabase) {
                super.onCreate(db)
                transferSongsInDB()
            }
        }

        private fun transferSongsInDB() {
            val tinydb = TinyDB(context)
            val p = PrefManager(context!!)
            val fav = context!!.getString(R.string.favTracks)

            val list = ArrayList<String>()

            CoroutineScope(Dispatchers.IO).launch {
                list.add(fav)
                val differ = async {
                    p.fetchAllPlayList().value!!
                }
                list.addAll(differ.await())
                for (pName in list) {
                    val songList = ArrayList<Song>()
                    songList.addAll(tinydb.getListObject(pName, Song::class.java))
                    for (song in songList) {
                        val playlist = Playlist(0, pName, song)
                        instance!!.playlistDao().addSong(playlist)
                    }
                }
            }
        }
    }
}